Core/include/Acts/TrackFitting/Chi2Fitter.hpp

```diff
@@ -0,0 +1,725 @@
+    using Chi2Result = typename Chi2Actor::result_type;
+    using Actors = ActionList<Chi2Actor>;
+    using Aborters = AbortList<Chi2Aborter>;
+
+    // Create relevant options for the propagation options
+    PropagatorOptions<Actors, Aborters> propOptions(
+        chi2FitterOptions.geoContext, chi2FitterOptions.magFieldContext,
+        logger);
+
+    // Set the trivial propagator options
+    propOptions.setPlainOptions(chi2FitterOptions.propagatorPlainOptions);
+
+    // Catch the actor and set the measurements
+    auto& chi2Actor = propOptions.actionList.template get<Chi2Actor>();
+    chi2Actor.inputMeasurements = &inputMeasurements;
+    chi2Actor.multipleScattering = chi2FitterOptions.multipleScattering;
+    chi2Actor.energyLoss = chi2FitterOptions.energyLoss;
+    chi2Actor.extensions = std::move(chi2FitterOptions.extensions);
+
+    using paramType = typename std::conditional<
+        std::is_same<start_parameters_t, parameters_t>::value,
+        std::variant<start_parameters_t>,
+        std::variant<start_parameters_t, parameters_t>>::type;
+    paramType vParams = sParameters;
+    // start_parameters_t and parameter_t can be the same
+
+    Chi2Result c2r;
+    // the result object which will be returned. Overriden every iteration.
+
+    for (int i = 0; i <= chi2FitterOptions.nUpdates; ++i) {
+      auto result = std::visit(
+          [mprop = m_propagator, propOptions](auto&& arg) {
+            return mprop.template propagate(arg, propOptions);
+          },
+          vParams);
+
+      if (!result.ok()) {
+        ACTS_ERROR("chi2 | it=" << i
+                                << " | propapation failed: " << result.error());
+        return result.error();
+      }
+
+      Chi2Result c2rCurrent = result.value().template get<Chi2Result>();
+
+      /// It could happen that the fit ends in zero measurement states.
+      /// The result gets meaningless so such case is regarded as fit failure.
+      if (c2rCurrent.result.ok() and not c2rCurrent.measurementStates) {
+        c2rCurrent.result = Result<void>(Chi2FitterError::NoMeasurementFound);
+      }
+
+      if (!c2rCurrent.result.ok()) {
+        ACTS_ERROR("chi2 | it=" << i << " | Chi2Fitter failed: "
+                                << c2rCurrent.result.error() << ", "
+                                << c2rCurrent.result.error().message());
+        return c2rCurrent.result.error();
+      }
+
+      c2rCurrent.residuals =
+          Eigen::Map<ActsDynamicVector>(c2rCurrent.collectorResiduals.data(),
+                                        c2rCurrent.collectorResiduals.size());
+      // TODO: is this safe? from Stackoverflow: "dangerous! Because the Eigen
+      // object will NOT create its own memory. It will operate on the memory
+      // provided by "data". In other words, working with the Eigen object
+      // when the "data" object is out of scope will result in a segmentation
+      // fault (or memory access violation)."
```

Comment:
I think this may be important to check or at least to catch it in some way.

Unfortunately I don't have expertise in that field.

Reply:
The comment is correct, if the lifetime of `c2rCurrent.residuals` is longer than `c2rCurrent.collectorResidual`, it will be a problem. Is this the case? Other options would be to

1. Create a copy into `.residuals` (so dropping the map type) which might be more expensive
2. Creating the map type "on demand" when it is used, and discarding it immediately after

I think I would probably go with option 2.